docs/public-api.md

```diff
@@ -0,0 +1,65 @@
+# SwiftNIO 2 Public API
+
+The SwiftNIO project (which includes all of `github.com/apple/swift-nio*`)aims to follow [SemVer 2.0.0](https://semver.org/spec/v2.0.0.html) which requires to declare a public API. This document along with the [API documentation](https://apple.github.io/swift-nio/) declare SwiftNIO's API.
+
+### What Are Acceptable Uses of NIO's Public API
+
+- it's acceptable and expected to use any type and call any method from the `NIO*` modules unless the type or function name start with an underscore (`_`).
+
+  What's the point of this restriction? If we prefix something with an underscore or put it into one of our internal modules, we can't commit to an API for it and don't expect users to ever use this. Also, there might be different underscored functions/types depending on the platform.
+  
+  Examples:
+   - ✅ `channel.close(promise: nil)`
+   - ❌ `channel._unsafe.flush0()`, underscored property
+   - ❌ `import CNIOAtomics`, module name doesn't start with NIO
+- it is acceptable to conform NIO types to protocols _you control_, ie. a protocol in _your codebase_. It's not acceptable to confirm any NIO types to protocols in the standard library or outside of your control.
+
+  What's the point of this restriction? NIO in a later new minor version might conform its type to said protocol.
+  
+  Examples:
+    - ✅ `extension EventLoopFuture: MyOwnProtocol { ... }`, assuming `MyOwnProtocol` lives in your codebase
+    - ❌ `extension EventLoopFuture: DebugStringConvertible { ... }`, `DebugStringConvertible` is a standard library protocol
+- it is acceptable to conform _your own types_ to NIO protocols but it's not acceptable to conform types you do not control to NIO protocols.
+
+  What's the point of this restriction? Again, NIO might later add this conformance.
+
+  Examples:
+    - ✅ `extension MyHandler: ChannelHandler { ... }`
+    - ❌ `extension Array: EventLoopGroup where Element: EventLoop { ... }`, as `Array` lives in the standard library
+
+- it is acceptable to add extensions to NIO types that either use one of your types as a non-default argument or are prefixed in a way that prevents clashes
```

Comment:
This one seems phrased a little on the restrictive side to me. Unlike in Objective-C, there's no run-time conflict when this rule is broken, and eventually we'll have a way to disambiguate.

If you wanted to split the difference, maybe this should be qualified to "public API"? Helpers within a module are generally acceptable (though I should double-check that we actually prefer them in tie-breakers).

The condition you're using may not be sufficient to cover everything, either, depending on how you phrase it. I could have a parameter of type `Array<MyType>`, but `[]` wouldn't uniquely identify it.

Reply:
@jrose-apple thank you! This was meant to be `public extensions`. This demonstrates the clash I mean:

(everything below a `// MODULE: xyz` would be in that module)
```
import BadModule
import RandomClient

doIt()
doItBad()

// MODULE: NIO
public struct ByteBuffer {
    public init() {
    }
    public func foo() {
        print("NIO.foo")
    }
}

// MODULE: BadModule
import NIO

public extension ByteBuffer {
    func foo() {
        print("BadModule.foo")
    }
}

public func doItBad() {
    ByteBuffer().foo()
}

// MODULE: RandomClient

import NIO
import BadModule

public func doIt() {
    ByteBuffer().foo()
}
```

which would get you

```
Compile Swift Module 'NIO' (1 sources)
Compile Swift Module 'BadModule' (1 sources)
Compile Swift Module 'RandomClient' (1 sources)
/private/tmp/.gen-swift-perf-test_vUBGJT/Sources/RandomClient/module.swift:35:1: error: ambiguous use of 'foo()'
ByteBuffer().foo()
^
NIO.ByteBuffer:3:17: note: found this candidate
    public func foo()
                ^
BadModule.ByteBuffer:2:17: note: found this candidate
    public func foo()
                ^
```

internal helpers are always fine here and they would be preferred which is what we want 👍

Reply:
Sorry, minor typo drive-by: s/confirm/conform/